src/modules/headset/oculus.c

```diff
@@ -349,6 +367,16 @@ static void oculus_renderTo(void (*callback)(void*), void* userdata) {
   double sensorSampleTime;
   getEyePoses(EyeRenderPose, &sensorSampleTime);
 
+  float delta = (float)(state.hapticLastTime - sensorSampleTime);
+  state.hapticLastTime = sensorSampleTime;
+  for (int i = 0; i < 2; ++i) {
+    ovr_SetControllerVibration(state.session, ovrControllerType_LTouch + i, state.hapticFrequency[i], state.hapticStrength[i]);
+    state.hapticDuration[i] -= delta;
+    if (state.hapticDuration[i] <= 0.0f) {
+      state.hapticStrength[i] = 0.0f;
+    }
+  }
+
```

Comment:
Maybe this belongs in `oculus_update`?  Feels more update-y than draw-y to me, but maybe here in renderTo makes more sense since you can access the predicted display time a bit more easily.  Not sure.

Reply:
oculus seems to want the controller updates to happen with frames and the timing information is already available, so it seemed most appropriate to put it here (despite being a bit non-render-y). I can move it if you want, but this is a bit simpler.